lorax/lorax.py

```diff
@@ -33,24 +34,340 @@ class TheLorax(object):
+                # Dropping the non-feature columns in the new test matrix, if they exist
+                # TODO: Handle the ID cols, Date cols elegantly
+                test_mat = test_mat.drop(dr_col, axis=1, errors='ignore')
+
+            if idx is not None:
+                sample = test_mat.loc[idx].values
+        
+        # Calculating Feature contributions
+        if isinstance(self.clf, RandomForestClassifier):
+            # Getting values for Random Forest Classifier
+            return_tuple = get_contrib_list_RF(self.clf, sample, self.column_names)
+
+            # Feature importance scores
+            contrib_list = return_tuple[4]
+        
+            # Model component information
+            self.model_info['num_trees'] = return_tuple[0]
+            self.model_info['global_score_dict'] = return_tuple[1]
+            self.model_info['feature_dict'] = return_tuple[2]
+            self.model_info['aggregated_dict'] = return_tuple[3]        
+            
+        elif isinstance(self.clf, LogisticRegression):
+            # Getting values for Random Forest Classifier
+            # TODO: The column names need to be consolidated
+
+            contrib_list = get_contrib_list_LR(self.clf, sample, self.column_names)
+
+        # Setting the prediction class
+        # predict proba returns a 2D array
+        scores = self.clf.predict_proba(sample.reshape(1, -1))[0]  
+
+        if pred_class is None:
+            # TODO: Multiclass adpatation
+            # use np.argmax(), or clf.predict()
+            pred_class = np.argmax(scores)
+
+        # Prediction score for the given pred_class
+        score = scores[pred_class]
+        
+        # TODO: handle this more elegantly for multiclass problems
+        # We need to flip the sign of the scores.
+        if pred_class == 0:
+            score = 1.0 - score
+            contrib_list = [(feature, score * -1) for feature, score in contrib_list]   
+
+        # TODO: Need to be modified to not taking the index
+        # Replacing the example id with -1 for now
+        logging.info('Used predicted class {} for example {}, score={}'.format(pred_class,
+                                                                            -1,
+                                                                            score))
+
+        # sorting in descending order by contribution then by feature name in the case of ties
+        contrib_list.sort(key=lambda x: (x[1] * -1, x[0]))
+
+        # TODO: If descriptive is set, the importance scores
+        # are supported with the context provided by a test dataset
+        # The code is available in the original constructor, move it here
+        if descriptive:  
+            # If descriptive, it rolls back to the original case         
+            if test_mat is None:
+                test_mat = self.X_test
+                fstats = self.feature_stats
+            else:
+                fstats = self.populate_feature_stats(test_mat)
```

Comment:
It does feel like it would be easy for users to fall into a pattern where they end up doing a lot of extra work if they repeatedly invoke `explain_example()` with the same test matrix (rather than registering it first), but the only potentially better option I see is if we maybe compared the passed matrix with what's already registered and then calculated the states only if they're different (actually, probably within `populate_feature_stats()` rather than here anyway). However, I'm not too sure how efficient comparing pandas matrices is whether its worth the overhead to avoid other calculations. Maybe add as an issue to look into sometime down the road?